zkevm-circuits/src/evm_circuit/op_execution/arithmetic.rs

```diff
@@ -0,0 +1,577 @@
+use super::super::{
+    BusMappingLookup, Case, Cell, Constraint, CoreStateInstance, ExecutionStep,
+    Lookup, Word,
+};
+use super::{CaseAllocation, CaseConfig, OpExecutionState, OpGadget};
+use bus_mapping::evm::OpcodeId;
+use halo2::{
+    arithmetic::FieldExt,
+    circuit::Region,
+    plonk::{Error, Expression},
+};
+use std::{array, convert::TryInto};
+
+#[derive(Clone, Debug)]
+struct AddSuccessAllocation<F> {
+    selector: Cell<F>,
+    swap: Cell<F>,
+    a: Word<F>,
+    b: Word<F>,
+    c: Word<F>,
+    carry: [Cell<F>; 32],
+}
+
+#[derive(Clone, Debug)]
+pub struct AddGadget<F> {
+    success: AddSuccessAllocation<F>,
+    stack_underflow: Cell<F>, // case selector
+    out_of_gas: (
+        Cell<F>, // case selector
+        Cell<F>, // gas available
+    ),
+}
+
+impl<F: FieldExt> OpGadget<F> for AddGadget<F> {
+    // AddGadget verifies ADD and SUB at the same time by an extra swap flag,
+    // when it's ADD, we annotate stack as [a, b, ...] and [c, ...],
+    // when it's SUB, we annotate stack as [a, c, ...] and [b, ...].
+    // Then we verify if a + b is equal to c.
+    const RESPONSIBLE_OPCODES: &'static [OpcodeId] =
+        &[OpcodeId::ADD, OpcodeId::SUB];
+
+    const CASE_CONFIGS: &'static [CaseConfig] = &[
+        CaseConfig {
+            case: Case::Success,
+            num_word: 3,  // a + b + c
+            num_cell: 33, // 32 carry + swap
+            will_halt: false,
+        },
+        CaseConfig {
+            case: Case::StackUnderflow,
+            num_word: 0,
+            num_cell: 0,
+            will_halt: true,
+        },
+        CaseConfig {
+            case: Case::OutOfGas,
+            num_word: 0,
+            num_cell: 0,
+            will_halt: true,
+        },
+    ];
+
+    fn construct(case_allocations: Vec<CaseAllocation<F>>) -> Self {
+        let [mut success, stack_underflow, out_of_gas]: [CaseAllocation<F>; 3] =
+            case_allocations.try_into().unwrap();
+        Self {
+            success: AddSuccessAllocation {
+                selector: success.selector,
+                swap: success.cells.pop().unwrap(),
+                a: success.words.pop().unwrap(),
+                b: success.words.pop().unwrap(),
+                c: success.words.pop().unwrap(),
+                carry: success.cells.try_into().unwrap(),
+            },
+            stack_underflow: stack_underflow.selector,
+            out_of_gas: (
+                out_of_gas.selector,
+                out_of_gas.resumption.unwrap().gas_available,
+            ),
+        }
+    }
+
+    fn constraints(
+        &self,
+        state_curr: &OpExecutionState<F>,
+        state_next: &OpExecutionState<F>,
+    ) -> Vec<Constraint<F>> {
+        let (add, sub) = (
+            Expression::Constant(F::from_u64(1)),
```

Comment:
use OpcodeId::ADD , OpcodeId::Sub here to replace hard code?

Reply:
Refactored in [`f0bd8af`](https://github.com/appliedzkp/zkevm-circuits/pull/26/commits/f0bd8afe2607abfb428c5fb171730625448ddd25). Thanks!